### ISSUE TITLE
OOM in validation loop

### ISSUE THREAD

**Issue by vladyorsh** [RF5/simple-sashimi#2]
Hi @RF5, thanks for the repo, it was of a big educational value for me. Reporting a small but annoying bug which makes it require >80GB machine to run.

The validation loop accumulates logits and labels before computing accuracy, which is a bit excessive (there is no pad masking anyways, so it's still somewhat off) and OOMs a 50GB RAM environment. Furthermore, it doesn't detach logits and accumulated loss from a computational graph, which leads to retaining them in memory. An example solution is:

```
if steps % cfg.validation_interval == 0 and steps != 0:
    model.eval()
    loss_fn.eval()
    torch.cuda.empty_cache()
    val_err_tot = 0
    acc_tot = 0
    with torch.no_grad():
        #Fastai pbars don't work for me, so I replaced them with tqdm
        for j, batch in tqdm(enumerate(validation_loader), total=len(validation_loader), leave=True, mininterval=5):
            x, y, lens = batch
            y = y.to(device)
            lens = lens.to(device)
            logits = model(x.to(device), lens)
            logits = logits.view(-1, cfg.model_cfg.mu_levels)
            y_ = y.view(-1)
            val_err_tot += loss_fn(logits, y_).detach() #Detach loss when accumulating
            acc_tot += (logits.argmax(dim=-1) == y_).float().mean().detach() #Accumulate hits

        val_err = val_err_tot / (j+1)
        acc     = acc_tot / (j + 1)
        sw.add_scalar('validation/acc', float(acc), steps)
        sw.add_scalar("validation/loss", val_err, steps)
        mb.write(f"validation run complete at {steps:,d} steps. validation loss: {val_err:5.4f}")
        # trust we must do this to stop memory leaks
        gc.collect()
```

**Comment by RF5** [comment]
Thanks for the tip! I made some slight changes which should partially address this; on the machine i evaluated on it had over 60GB RAM so I did not encounter this issue.
